src/vs/editor/common/model/textModelTokens.ts

```diff
@@ -208,102 +221,74 @@ export class TextModelTokenization extends Disposable {
-		// if the `deadline` object will be valid after execution leaves this function.
-		const endTime = Date.now() + deadline.timeRemaining();
-
-		const execute = () => {
-			if (this._isDisposed || !this._textModel.isAttachedToEditor() || !this._hasLinesToTokenize()) {
-				// disposed in the meantime or detached or finished
-				return;
-			}
 
-			this._backgroundTokenizeForAtLeast1ms();
+		this.backgroundTokenizer.clear();
 
-			if (Date.now() < endTime) {
-				// There is still time before reaching the deadline, so yield to the browser and then
-				// continue execution
-				setTimeout0(execute);
-			} else {
-				// The deadline has been reached, so schedule a new idle callback if necessary
-				this._beginBackgroundTokenization();
+		this._defaultBackgroundTokenizer = null;
+		if (this._tokenizationStateStore) {
+			const b: IBackgroundTokenizationStore = {
+				setTokens: (tokens) => {
+					this._tokenizationPart.setTokens(tokens);
+				},
+				backgroundTokenizationFinished: () => {
+					this._tokenizationPart.handleBackgroundTokenizationFinished();
+				},
+				setEndState: (lineNumber, state) => {
+					if (!state) {
+						throw new BugIndicatingError();
+					}
+					const invalidLineStartIndex = this._tokenizationStateStore?.invalidLineStartIndex;
+					if (invalidLineStartIndex !== undefined && lineNumber - 1 >= invalidLineStartIndex) {
+						// Don't accept states for definitely valid states
+						this._tokenizationStateStore?.setEndState(this._textModel.getLineCount(), lineNumber - 1, state);
+					}
+				},
+			};
+
+			this.backgroundTokenizer.clear();
+
+			if (tokenizationSupport && tokenizationSupport.createBackgroundTokenizer) {
+				this.backgroundTokenizer.value = tokenizationSupport.createBackgroundTokenizer(this._textModel, b);
 			}
-		};
-		execute();
-	}
-
-	/**
-	 * Tokenize for at least 1ms.
-	 */
-	private _backgroundTokenizeForAtLeast1ms(): void {
-		const lineCount = this._textModel.getLineCount();
-		const builder = new ContiguousMultilineTokensBuilder();
-		const sw = StopWatch.create(false);
-
-		do {
-			if (sw.elapsed() > 1) {
-				// the comparison is intentionally > 1 and not >= 1 to ensure that
-				// a full millisecond has elapsed, given how microseconds are rounded
-				// to milliseconds
-				break;
+			if (!this.backgroundTokenizer.value) {
+				this.backgroundTokenizer.value = this._defaultBackgroundTokenizer =
+					new DefaultBackgroundTokenizer(
+						this._textModel,
+						this._tokenizationStateStore!,
+						b,
+						this._languageIdCodec
+					);
+				this._defaultBackgroundTokenizer.handleChanges();
 			}
-
-			const tokenizedLineNumber = this._tokenizeOneInvalidLine(builder);
-
-			if (tokenizedLineNumber >= lineCount) {
-				break;
-			}
-		} while (this._hasLinesToTokenize());
-
-		this._tokenizationPart.setTokens(builder.finalize(), this._isTokenizationComplete());
+		} else {
+			this.backgroundTokenizer.clear();
```

Comment:
this branch is not necessary because `this.backgroundTokenizer.clear()` is called before.